src/aws-cpp-sdk-core/include/aws/core/utils/stream/AwsChunkedStream.h

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#pragma once
+#include <aws/core/utils/HashingUtils.h>
+#include <aws/core/utils/Outcome.h>
+#include <aws/core/utils/crypto/Hash.h>
+
+namespace Aws {
+namespace Utils {
+namespace Stream {
+
+static const size_t AWS_DATA_BUFFER_SIZE = 65536;
+
+template <size_t DataBufferSize = AWS_DATA_BUFFER_SIZE>
+class AwsChunkedStream {
+ public:
+  AwsChunkedStream(Http::HttpRequest *request, const std::shared_ptr<Aws::IOStream> &stream)
+      : m_chunkingStream{Aws::MakeShared<StringStream>("AwsChunkedStream")}, m_request(request), m_stream(stream) {
+    assert(m_stream != nullptr);
+    if (m_stream == nullptr) {
+      AWS_LOGSTREAM_ERROR("AwsChunkedStream", "stream is null");
+    }
+    assert(request != nullptr);
+    if (request == nullptr) {
+      AWS_LOGSTREAM_ERROR("AwsChunkedStream", "request is null");
+    }
+  }
+
+  size_t BufferedRead(char *dst, size_t amountToRead) {
+    assert(dst != nullptr);
+    if (dst == nullptr) {
+      AWS_LOGSTREAM_ERROR("AwsChunkedStream", "dst is null");
+    }
+
+    // the chunk has ended and cannot be read from
+    if (m_chunkEnd) {
+      return 0;
+    }
+
+    // If we've read all of the underlying stream write the checksum trailing header
+    // the set that the chunked stream is over.
+    if (m_stream->eof() && !m_stream->bad() && (m_chunkingStream->eof() || m_chunkingStream->peek() == EOF)) {
+      Aws::StringStream chunkedTrailerStream;
+      chunkedTrailerStream << "0\r\n";
+      if (m_request->GetRequestHash().second != nullptr) {
+        chunkedTrailerStream << "x-amz-checksum-" << m_request->GetRequestHash().first << ":"
+                             << HashingUtils::Base64Encode(m_request->GetRequestHash().second->GetHash().GetResult()) << "\r\n";
+      }
+      chunkedTrailerStream << "\r\n";
+      const auto chunkedTrailer = chunkedTrailerStream.str();
+      auto trailerSize = chunkedTrailer.size();
+      assert(amountToRead >= trailerSize);
+      memcpy(dst, chunkedTrailer.c_str(), std::min(trailerSize, amountToRead));
+      m_chunkEnd = true;
+      return trailerSize;
+    }
+
+    // Try to read in a 64K chunk, if we cant we know the stream is over
+    size_t bytesRead = 0;
+    while (m_stream->good() && bytesRead < DataBufferSize) {
```

Comment:
why not bytesRead <= DataBufferSize ?

Reply:
that will actually trigger an [assertion here](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/include/aws/core/utils/Array.h#L206-L210), m_data is of size `DataBufferSize`, its created using this [constructor](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/include/aws/core/utils/Array.h#L42-L47) in our "dynamic array" class. its not a real array